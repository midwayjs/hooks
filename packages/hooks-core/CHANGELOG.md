# @midwayjs/hooks-core

## 3.0.0-rc.3

### Patch Changes

- Set correct version for examples

## 3.0.0-rc.2

### Patch Changes

- Build after publish

## 3.0.0-rc.1

### Patch Changes

- 71a3115: Release 3.0.0 rc version
