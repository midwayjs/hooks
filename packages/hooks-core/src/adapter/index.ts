import { createDebug, validateFunction } from '../common'
import { ResponseMetaData } from '../api/operator/http'

const debug = createDebug('hooks-core: adapter')

type FrameworkAdapter = {
  name: string
  handleResponseMetaData(metadata: ResponseMetaData[]): Promise<any>
}

export let framework: FrameworkAdapter = null

export function useFrameworkAdapter(adapter: FrameworkAdapter) {
  if (framework) {
    throw new Error(
      `FrameworkAdapter is already set, current: ${framework.name}, new: ${adapter.name}`
    )
  }
  validateFunction(
    adapter.handleResponseMetaData,
    'adapter.handleResponseMetaData'
  )
  debug('use framework adapter: %s', adapter.name)
  framework = adapter
}
