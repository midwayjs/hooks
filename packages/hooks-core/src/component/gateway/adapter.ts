import {
  IMidwayApplication,
  IMidwayContainer,
  IMidwayContext,
} from '@midwayjs/core'
import { ApiFunction, ServerRoute } from '../..'
import { ComponentConfig } from './interface'

export interface CreateApiParam {
  fn: ApiFunction
  file: string
  isExportDefault: boolean
}

export interface HooksGatewayAdapter {
  config: ComponentConfig
  container: IMidwayContainer
  app: IMidwayApplication<IMidwayContext>

  createApi(config: CreateApiParam): void
  is(route: ServerRoute): boolean
  afterCreate?(): void

  onError(ctx: any, error: any): void

  getGlobalMiddleware?(): any[]
}
