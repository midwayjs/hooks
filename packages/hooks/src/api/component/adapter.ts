import {
  AbstractRouter,
  ApiRoute,
  ContextManager,
  createDebug,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  HooksMiddleware,
  HttpTrigger,
  ResponseMetaData,
  ResponseMetaType,
  validateOneOf,
} from '@midwayjs/hooks-core'
import type { IMidwayApplication, IMidwayContainer } from '@midwayjs/core'
import {
  MidwayApplicationManager,
  MidwayServerlessFunctionService,
  MidwayWebRouterService,
  All,
  Del,
  Get,
  Head,
  Options,
  Patch,
  Post,
  Put,
  ServerlessTriggerType,
} from '@midwayjs/core'
import { FileSystemRouter, normalizePath } from '@midwayjs/hooks-internal'
import camelCase from 'lodash/camelCase'
import { ServerlessTrigger } from '../operator/serverless'
import { useHooksMiddleware } from '../middleware'
import { useContext } from '../hooks'

const debug = createDebug('hooks: MidwayFrameworkAdapter')

export interface MidwayApplication extends IMidwayApplication {
  use?: (middleware: any) => void
}

export class MidwayFrameworkAdapter {
  constructor(
    public source: string,
    public router: AbstractRouter,
    public app: MidwayApplication
  ) {}

  async initialize(
    container: IMidwayContainer,
    apis: ApiRoute[],
    globalMiddlewares: HooksMiddleware[]
  ) {
    const applicationManager = await container.getAsync(
      MidwayApplicationManager
    )

    const faas = applicationManager.getApplication('faas')
    const koa = applicationManager.getApplication('koa')

    if (faas) {
      this.app = faas
      this.createServerlessApi(
        apis,
        await container.getAsync(MidwayServerlessFunctionService)
      )
    } else {
      this.app = koa
      this.createWebApi(apis, await container.getAsync(MidwayWebRouterService))
    }

    this.registerGlobalMiddleware(globalMiddlewares)
  }

  isHttpTrigger(api: ApiRoute): api is ApiRoute<HttpTrigger> {
    return api.trigger.type === 'HTTP'
  }

  isServerlessTrigger(api: ApiRoute): api is ApiRoute<ServerlessTrigger> {
    return typeof (api.trigger as ServerlessTrigger)?.parseArgs === 'function'
  }

  createWebApi(
    apis: ApiRoute[],
    midwayWebRouterService: MidwayWebRouterService
  ) {
    for (const api of apis) {
      const providerId = this.getUniqueProviderId(api)

      debug('create trigger: %s, providerId: %s', api.trigger.type, providerId)

      if (this.isHttpTrigger(api)) {
        const http = this.createHttpApi(api)
        midwayWebRouterService.addRouter(http.handler, {
          url: http.path,
          middleware: http.middleware,
          requestMethod: http.method,
        })
      }
    }
  }

  createHttpApi(api: ApiRoute<HttpTrigger>) {
    validateOneOf(
      api.trigger.method,
      'trigger.method',
      Object.keys(this.methodDecorators)
    )

    const parseHTTPArgs = ({ ctx }) => {
      return ctx.request?.body?.args || []
    }

    return {
      path: normalizePath(this.router, api),
      middleware: api.middleware?.map(useHooksMiddleware),
      handler: this.createHandler(api.fn, parseHTTPArgs),
      method: api.trigger.method.toLowerCase(),
    }
  }

  private createHandler(
    fn: Function,
    parseArgs: (inputs: { ctx: any; args: any[] }) => any[]
  ) {
    return (ctx: any, ...args: unknown[]) => {
      return fn(...parseArgs({ ctx, args }))
    }
  }

  createServerlessApi(
    apis: ApiRoute[],
    midwayServerlessFunctionService: MidwayServerlessFunctionService
  ) {
    for (const api of apis) {
      const providerId = this.getUniqueProviderId(api)

      debug('create trigger: %s, providerId: %s', api.trigger.type, providerId)

      if (this.isHttpTrigger(api)) {
        const http = this.createHttpApi(api)
        midwayServerlessFunctionService.addServerlessFunction(http.handler, {
          type: ServerlessTriggerType.HTTP,
          metadata: {
            name: ServerlessTriggerType.HTTP,
            method: http.method as any,
            path: http.path,
            middleware: http.middleware,
          },
          functionName: providerId,
          handlerName: `${providerId}.handler`,
        })
        continue
      }

      if (this.isServerlessTrigger(api)) {
        midwayServerlessFunctionService.addServerlessFunction(
          this.createHandler(api.fn, api.trigger.parseArgs),
          {
            name: api.trigger.type,
            type: api.trigger.type,
            metadata: api.trigger.options,
            functionName: providerId,
            handlerName: `${providerId}.handler`,
          } as any
        )
      }
    }
  }

  private getUniqueProviderId(api: ApiRoute) {
    if (this.router instanceof FileSystemRouter) return api.functionId
    if (api.functionName !== EXPORT_DEFAULT_FUNCTION_ALIAS)
      return camelCase(api.functionId)

    // api router & export default function
    const router = new FileSystemRouter({
      source: this.source,
      routes: [],
    })
    return router.getFunctionId(api.file, api.functionName, true)
  }

  private methodDecorators = {
    GET: Get,
    POST: Post,
    PUT: Put,
    DELETE: Del,
    PATCH: Patch,
    HEAD: Head,
    OPTIONS: Options,
    ALL: All,
  }

  registerGlobalMiddleware(middlewares: HooksMiddleware[] = []) {
    for (const mw of middlewares) {
      this.app.use?.(useHooksMiddleware(mw))
    }
  }

  private async useUniversalRuntime(ctx: any, next: any) {
    return await ContextManager.run({ ctx }, async () => await next())
  }

  async handleResponseMetaData(metadata: ResponseMetaData[]): Promise<any> {
    const ctx = useContext()

    for (const meta of metadata) {
      switch (meta.type) {
        case ResponseMetaType.CODE:
          ctx.status = meta.code
          break
        case ResponseMetaType.HEADER:
          ctx.set(meta.header.key, meta.header.value)
          break
        case ResponseMetaType.CONTENT_TYPE:
          ctx.type = meta.contentType
          break
        case ResponseMetaType.REDIRECT:
          ctx.status = meta.code || 302
          ctx.redirect(meta.url)
          break
      }
    }
  }
}
