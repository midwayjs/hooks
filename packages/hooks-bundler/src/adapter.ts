import { AbstractRouter, ApiRoute, HttpTriggerType } from '@midwayjs/hooks-core'
import { AbstractBundlerAdapter, requireWithoutCache } from '@midwayjs/bundler'
import { join } from 'upath'
import { createExpressDevPack } from '@midwayjs/dev-pack'
import {
  getRouter,
  getSource,
  normalizeUrl,
  ProjectConfig,
} from '@midwayjs/hooks/internal'
import { getConfig, getProjectRoot } from '@midwayjs/hooks-config'
import type { ViteDevServer } from 'vite'
import cloneDeep from 'lodash/cloneDeep'

export class MidwayBundlerAdapter extends AbstractBundlerAdapter {
  protected root: string
  protected source: string
  protected router: AbstractRouter

  protected projectConfig: ProjectConfig

  constructor() {
    const source = getSource({ useSourceFile: true })
    const router = getRouter(source)
    super({
      name: '@midwayjs/hooks-bundler',
      router,
    })
    this.root = getProjectRoot()
    this.projectConfig = getConfig()
    this.source = source
    this.router = router
  }

  override transformApiRoutes(apis: ApiRoute[]): ApiRoute[] {
    apis = cloneDeep(apis)
    for (const api of apis) {
      if (api.trigger.type === HttpTriggerType) {
        api.trigger.path = normalizeUrl(this.router, api)
      }
    }
    return apis
  }

  override getSource(): string {
    return this.source
  }

  override getUnplugin() {
    const ctx = this
    return {
      vite: {
        resolveId(_: string, importer: string) {
          if (
            process.env.NODE_ENV !== 'production' &&
            importer &&
            ctx.router.isSourceFile(importer, ctx.source) &&
            ctx.router.isApiFile({
              mod: requireWithoutCache(importer),
              file: importer,
            })
          ) {
            return 'MIDWAY_HOOKS_VIRTUAL_FILE'
          }

          return null
        },
        async configureServer(server: ViteDevServer) {
          const { middleware } = await createExpressDevPack({
            cwd: ctx.root,
            sourceDir: join(ctx.root, ctx.projectConfig.source),
            watch: true,
          })

          server.middlewares.use(middleware)
        },
        config() {
          return {
            optimizeDeps: {
              include: ['@midwayjs/rpc'],
            },
            build: {
              manifest: true,
            },
          }
        },
      },
    }
  }
}
