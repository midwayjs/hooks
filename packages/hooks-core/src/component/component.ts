import parseArgs from 'fn-args'
import { sync } from 'globby'
import _ from 'lodash'
import path from 'path'
import upath from 'upath'

import { createConfiguration, IMidwayContainer } from '@midwayjs/core'

import { ServerRoute } from '..'
import { als } from '../runtime'
import { ApiFunction, ApiModule } from '../types/common'
import { ComponentConfig, HooksGatewayAdapter } from '../types/gateway'
import { ApiHttpMethod } from '../types/http'
import { useHooksMiddleware } from '../util'
import { HTTPGateway } from './gateway/http'

export class HooksComponent {
  config: ComponentConfig

  adapters: HooksGatewayAdapter[]

  constructor(config: ComponentConfig) {
    this.config = config
    this.adapters = config.runtime.gatewayAdapters.map(
      (adapter) => new adapter(this.config)
    )
  }

  init() {
    const {
      router,
      internal: { routes },
    } = this.config

    let count = 0
    const totalCount = routes.reduce((totalCount, route) => {
      // Windows
      const dir = upath.join(router.source, route.baseDir)
      const files = sync([dir]).filter((file) => router.isApiFile(file))
      return totalCount + files.length
    }, 0)

    const configuration = createConfiguration({
      namespace: '@midwayjs/hooks',
      directoryResolveFilter: routes.map((route) => {
        return {
          // Windows
          pattern: path.join(router.source, route.baseDir),
          ignoreRequire: true,
          filter: (_: void, file: string, container: IMidwayContainer) => {
            if (!router.isApiFile(file)) {
              return
            }

            // Initialize container
            this.adapters.forEach((adapter) => (adapter.container = container))

            // Create api function
            const adapter = this.getAdapterByRoute(route)
            const mod: ApiModule = require(file)
            this.createApi(mod, file, adapter)
            container.bindClass(mod, '', file)

            // Call afterCreate hooks
            count++
            if (count === totalCount) {
              this.adapters.forEach((adapter) => adapter.afterCreate?.())
            }
          },
        }
      }),
    })

    configuration
      .onReady((_, app) => {
        this.adapters.forEach((adapter) => (adapter.app = app))

        // Setup global middleware
        for (const mw of this.getGlobalMiddleware()) {
          ;(app as any).use(mw)
        }
      })
      .onStop(_.noop)

    return {
      Configuration: configuration,
    }
  }

  getAdapterByRoute(route: ServerRoute<any>) {
    const adapter = this.adapters.find((adapter) => adapter.is(route))

    if (!adapter) {
      throw new Error(
        `Can't find the correct gateway adapter, please check if midway.config.ts is correct`
      )
    }

    return adapter
  }

  createApi(mod: ApiModule, file: string, adapter: HooksGatewayAdapter) {
    const modMiddleware = mod?.config?.middleware || []
    const funcs = _.pickBy<ApiFunction>(mod, _.isFunction)

    for (const [name, fn] of Object.entries(funcs)) {
      fn.middleware = (fn.middleware || (fn.middleware = []))
        .concat(modMiddleware)
        .map(useHooksMiddleware)

      const isExportDefault = name === 'default'
      const functionName = isExportDefault ? '$default' : name
      const id = this.config.router.getFunctionId(
        file,
        functionName,
        isExportDefault
      )

      let httpPath = ''

      if (adapter instanceof HTTPGateway) {
        httpPath = this.config.router.getHTTPPath(
          file,
          functionName,
          isExportDefault
        )
      }

      const httpMethod: ApiHttpMethod =
        parseArgs(fn).length === 0 ? 'GET' : 'POST'

      // Set param for unit testing
      fn._param = {
        url: httpPath,
        method: httpMethod,
        meta: { functionName: id },
      }

      adapter.createApi({ fn, id, httpPath })
    }
  }

  getGlobalMiddleware() {
    const mws = [this.useAsyncLocalStorage]
    this.config.runtime.middleware?.forEach?.((mw) =>
      mws.push(useHooksMiddleware(mw))
    )
    return mws
  }

  useAsyncLocalStorage = async (ctx: any, next: any) => {
    await als.run({ ctx }, async () => {
      await next()
    })
  }
}
