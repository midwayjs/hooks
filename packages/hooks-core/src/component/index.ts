import { ServerRouter } from '../router'
import { getConfig, getProjectRoot } from '../config'
import { RuntimeConfig } from '../types/config'
import { isDevelopment } from '../util'
import { HooksComponent } from './component'
import { HTTPGateway } from './gateway/http'

export { HTTPGateway } from './gateway/http'

/**
 * Create hooks component
 */
export const hooks = (runtime: RuntimeConfig = {}) => {
  runtime.adapter = runtime.adapter || HTTPGateway

  const root = getProjectRoot()
  const internal = getConfig()
  const router = new ServerRouter(root, internal, isDevelopment())

  const cmp = new HooksComponent({
    runtime,
    internal,
    router,
    root,
  })

  return cmp.init()
}
