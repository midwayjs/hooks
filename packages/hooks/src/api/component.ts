import { IMidwayContainer } from '@midwayjs/core'
import { ApplicationContext, Configuration, Init } from '@midwayjs/decorator'
import { __decorate } from 'tslib'
import {
  AbstractRouter,
  ApiRouter,
  createDebug,
  useFrameworkAdapter,
  validateArray,
} from '@midwayjs/hooks-core'
import {
  getHydrateOptions,
  getRouter,
  getSource,
  HOOKS_DEV_MODULE_PATH,
  isDev,
  isHydrate,
  loadApiFiles,
  parseApiModule,
  RuntimeConfig,
} from '@midwayjs/hooks-internal'
import flattenDeep from 'lodash/flattenDeep'
import { MidwayFrameworkAdapter } from './component/adapter'

const debug = createDebug('hooks: component')

export function HooksComponent(runtimeConfig: RuntimeConfig = {}) {
  if (runtimeConfig.middleware !== undefined) {
    validateArray(runtimeConfig.middleware, 'runtimeConfig.middleware')
  }

  const useSourceFile = isDev()
  debug('useSourceFile: %s', useSourceFile)

  const source = getSource({ useSourceFile })
  const router = getRouter(source)
  const midway = new MidwayFrameworkAdapter(source, router, null)

  useFrameworkAdapter({
    name: 'Midway',
    handleResponseMetaData: midway.handleResponseMetaData,
  })

  const apis = isHydrate()
    ? loadHydrateModules(router)
    : loadApiModules(source, router)

  if (apis.length === 0) {
    console.warn('No api routes found, source is:', source)
  }

  debug('HOOKS_DEV_MODULE_PATH', process.env[HOOKS_DEV_MODULE_PATH])
  if (process.env[HOOKS_DEV_MODULE_PATH]) {
    const devApis = loadApiModules(
      process.env[HOOKS_DEV_MODULE_PATH],
      router instanceof ApiRouter ? router : new ApiRouter()
    )
    debug('load dev apis %O', devApis)
    apis.push(...devApis)
  }

  const Configuration = createConfiguration(async (container) => {
    await midway.initialize(container, apis, runtimeConfig.middleware)
  })

  return { Configuration }
}

// source: https://www.typescriptlang.org/play?target=99&moduleResolution=99&importHelpers=true&emitDecoratorMetadata=false&pretty=true#code/JYWwDg9gTgLgBAbzgSQHbBgGjgQTGAG2AGMBDGYCVAYSpgFMAPLOW1AM2AHMBXKcyqjgBfOOygQQcAEQABEMAAmAd1IBPAFYBnAPSL6xaOWjSA3AChL52W069+FKgAoE5uHFSkQ9LWFLF6AC4ZeSVVTV0ACwgIAGstaUxzYQBKc2ICUi0tOAAJGPjbbgdBRDc4WTxCEgEqNgZmJzT3Q1QYUmBUeihg0lQ1C3LZNAwm8qy1VGI4TtGUsvdFuFJVDDgqEZgnGEjgLQA6VvbO7uaRZPMgA
function createConfiguration(
  onInit: (container: IMidwayContainer) => Promise<void>
) {
  let HooksConfiguration = class HooksConfiguration {
    container: any
    async init() {
      await onInit(this.container)
    }
  }
  __decorate(
    [ApplicationContext()],
    HooksConfiguration.prototype,
    'container',
    void 0
  )
  __decorate([Init()], HooksConfiguration.prototype, 'init', null)
  HooksConfiguration = __decorate(
    [
      Configuration({
        namespace: '@midwayjs/hooks',
        importConfigs: [
          {
            default: {
              asyncContextManager: {
                enable: true,
              },
            },
          },
        ],
      }),
    ],
    HooksConfiguration
  )

  return HooksConfiguration
}

function loadHydrateModules(router: AbstractRouter) {
  const { modules } = getHydrateOptions()
  debug('load hydrate modules %O', modules)
  const routes = modules.map(({ mod, file }) =>
    parseApiModule(mod, file, router)
  )
  return flattenDeep(routes)
}

function loadApiModules(source: string, router: AbstractRouter) {
  const { apis } = loadApiFiles({
    source,
    router,
  })
  const routes = apis.map((file) => parseApiModule(require(file), file, router))
  return flattenDeep(routes)
}
