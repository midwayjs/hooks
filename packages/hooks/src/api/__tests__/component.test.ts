import {
  CONTROLLER_KEY,
  getClassMetadata,
  listModule,
  WEB_ROUTER_KEY,
} from '@midwayjs/decorator'
import { HttpMethod } from '@midwayjs/hooks-core'
import { HooksComponent } from '../component'
import { MidwayFrameworkAdapter } from '../component/adapter'

describe('component', () => {
  const framework = new MidwayFrameworkAdapter(null, null, null, null)

  it('component should validate arguments', () => {
    expect(() =>
      HooksComponent({ middleware: null })
    ).toThrowErrorMatchingSnapshot()

    expect(() =>
      framework.createHttpApi({
        trigger: {
          type: 'HTTP',
          method: 'ERROR',
        },
      } as any)
    ).toThrowErrorMatchingSnapshot()
  })

  it('should register api route as Controller', async () => {
    framework.createHttpApi({
      fn: async () => {},
      file: '/index.ts',
      functionName: 'index',
      trigger: {
        type: 'HTTP',
        method: HttpMethod.GET,
        path: '/',
      },
      middleware: [],
      functionId: 'api-index',
    })

    framework.createHttpApi({
      fn: async () => {},
      file: '/index.ts',
      functionName: 'post',
      trigger: {
        type: 'HTTP',
        method: HttpMethod.POST,
        path: '/post',
      },
      middleware: [async () => {}],
      functionId: 'api-index-post',
    })

    const Controllers = listModule(CONTROLLER_KEY)
    expect(Controllers.length).toBe(2)

    for (const Controller of Controllers) {
      const ControllerRoute = getClassMetadata(WEB_ROUTER_KEY, Controller)
      const ControllerMeta = getClassMetadata(CONTROLLER_KEY, Controller)

      expect({
        route: ControllerRoute,
        controller: ControllerMeta,
      }).toMatchSnapshot()
    }
  })
})
