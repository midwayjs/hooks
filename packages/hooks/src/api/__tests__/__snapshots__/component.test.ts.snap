// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component component should validate arguments 1`] = `"[ERR_INVALID_ARG_TYPE]: The 'runtimeConfig.middleware' argument must be of type Array. Received null"`;

exports[`component component should validate arguments 2`] = `"[ERR_INVALID_ARG_VALUE]: The argument 'ERROR' must be one of: 'GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'HEAD', 'OPTIONS', 'ALL' . Received type string ('trigger.method')"`;

exports[`component should register api route as Controller 1`] = `
{
  "handler": [Function],
  "method": "get",
  "middleware": [],
  "path": "/",
}
`;

exports[`component should register api route as Controller 2`] = `
{
  "handler": [Function],
  "method": "post",
  "middleware": [
    [Function],
  ],
  "path": "/post",
}
`;
